src/backend/d3d12/ShaderModuleD3D12.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ShaderModuleD3D12.h"
+
+#include <spirv-cross/spirv_hlsl.hpp>
+
+namespace backend {
+namespace d3d12 {
+
+    std::string GetBindingName(uint32_t group, uint32_t binding) {
+        std::ostringstream o;
+        o << "nxt_binding_" << group << "_" << binding;
+        return o.str();
+    }
+
+    bool operator < (const BindingLocation& a, const BindingLocation& b) {
+        return std::tie(a.group, a.binding) < std::tie(b.group, b.binding);
+    }
+
+    bool operator < (const CombinedSampler& a, const CombinedSampler& b) {
+        return std::tie(a.samplerLocation, a.textureLocation) < std::tie(b.samplerLocation, b.textureLocation);
+    }
+
+    std::string CombinedSampler::GetName() const {
+        std::ostringstream o;
+        o << "nxt_combined";
+        o << "_" << samplerLocation.group << "_" << samplerLocation.binding;
+        o << "_with_" << textureLocation.group << "_" << textureLocation.binding;
+        return o.str();
+    }
+
+    ShaderModule::ShaderModule(Device* device, ShaderModuleBuilder* builder)
+        : ShaderModuleBase(builder), device(device) {
+        spirv_cross::CompilerHLSL compiler(builder->AcquireSpirv());
+
+        spirv_cross::CompilerHLSL::Options options;
+        options.shader_model = 40; 
+        options.flip_vert_y = false;
+        options.fixup_clipspace = true;
+
+        compiler.set_options(options);
+
+        ExtractSpirvInfo(compiler);
+
+        const auto& bindingInfo = GetBindingInfo();
```

Comment:
I think you can safely remove everything below except hlslSource = ... This would allow dead-removing a whole lot of code.